[GEN][ZH] Implement System Time and Game Time display

Mauller opened the pull request:
**Squash and merge**
- Resolves #903
- Relates To: #850 

This PR implements the system clock and game time. They can be stylised and moved by adding the relevant fields to the inGameUi.ini file.

The system time is just currently offset from the left while the game time is positioned relative to the right side of the viewport.
This is to replicate how they are implemented within GenTool.

They are enabled by default for now to a font size of 8, but can also be disabled in the options.ini by setting their font size to 0.

The following will disable them independently if added to the options.ini.
```
SystemTimeFontSize = 0
GameTimeFontSize = 0
```

EDIT: Updated example with latest version
EDIT2: Updated once more with the now grey frame string
Example:
![image](https://github.com/user-attachments/assets/4afa635d-a046-44d1-9876-edb063f5a8b8)

## TODO

- [x] Test system time with -quickstart
- [x] Replicate in Generals
- [x] Update the Pull Request description

xezon commented:
The placement of the right side timer is not optimal.

What are the options of fonts here?

GenTool uses the "Tahoma" font for its top screen texts, starting with a size of 8. I do wonder if it would make sense to use that font again, to present it with a familiar look for legacy players.

xezon commented:
In GenTool I have a function to get the extent for font + text to get its width and height. With this information, refined positioning is possible.

Mauller commented:
Just a rebase with recent main before making relevant changes.

Mauller commented:
Updated with recent changes, the sim timer is not ahcnored relative to the right of the screen, so adjusting the x coordinate will shift it left instead of right.

Also updated to use Tohama as the default font at 8 point.

![image](https://github.com/user-attachments/assets/0388bc81-3233-4ec8-a9a3-bca6bc640ca4)

Mauller commented:
Updated with Tahoma set to bold and rebased with last 5 pushes to main.

Mauller commented:
Rebased with main and updated the colour of the frame string to be the same grey as Gentools

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/OptionsMenu.cpp:
```diff
@@ -772,6 +772,32 @@ Real OptionPreferences::getMusicVolume(void)
+{
+	OptionPreferences::const_iterator it = find("SystemTimeEnabled");
+	if (it == end())
+		return TheGlobalData->m_showSystemTime;
```
I prefer to not use GlobalData for the default values here. It is a strange setup because GlobalData calls this function.

Instead, return a Tri State: Undefined, True, False and then work with that.

Mauller replied:
I changed it to return true by default if no option is found from the Options.ini, this then sets the value in the global data for the relevant option to true instead of using the default initialised value in globaldata.

xezon replied:
Ok that is fine too.

xezon commented on GeneralsMD/Code/GameEngine/Include/Common/GlobalData.h:
```diff
@@ -124,6 +124,11 @@ class GlobalData : public SubsystemInterface
+	// TheSuperHAckers @feature Mauller 21/06/2025
+	// Allow the system time and sim timer to be toggleable
+	Bool m_showSystemTime;
+	Bool m_showSimulationTimer;
```
Can move these elsewhere to other similar settings, instead of putting between "headless" and "windowed", which are very unrelated.

Mauller replied:
i placed them where they are as they will fill padding bytes and not extend the size of the global data object.

xezon replied:
Maybe there are more suitable free pads elsewhere?

Mauller replied:
I moved it to another place where things are being toggled to be displayed, but theres not really any great places that has relevant data.

xezon commented on GeneralsMD/Code/GameEngine/Include/GameClient/InGameUI.h:
```diff
@@ -741,6 +742,25 @@ friend class Drawable;	// for selection/deselection transactions
+	Bool												m_systemTimeBold;
+	Coord2D												m_systemTimePosition;
+	Color												m_systemTimeColor;
+	Color												m_systemTimeDropColor;
```
Perhaps these 2 blocks are better combined as struct?

Like

```cpp
struct DisplaySomething
{
	DisplayString * string;
	AsciiString font;
	Int	pointSize;
	Bool bold;
	Coord2D position;
	Color color;
	Color dropColor;
};

DisplaySomething m_systemTime;
DisplaySomething m_gameTime;
```

Then DisplayString life time management can also be in the struct.

Mauller replied:
> Then DisplayString life time management can also be in the struct.

The display strings need to be cleaned up during ingameUI reset since the display string manager throws a wobbly if things are done out of order.

xezon commented on GeneralsMD/Code/GameEngine/Include/GameClient/InGameUI.h:
```diff
@@ -741,6 +742,25 @@ friend class Drawable;	// for selection/deselection transactions
+
+	// Simulation Timer
+	DisplayString *										m_simulationTimerString;
+	DisplayString *										m_simulationTimerFrameString;
```
Can we not reuse DisplayStrings here? Every single string instance needs to be unique?

Mauller replied:
> Can we not reuse DisplayStrings here? Every single string instance needs to be unique?

it seems like they can be reused, but in this instance since we don't have inline colour handling, this actually makes handling the strings simpler since i need to get the string width for the two seperate parts.

If i used a single string i would have to set the string multiple times to get the widths then draw it out which might look less intuitive on the surface compared to the two seperate strings being used in the way i currently have it being handled.

Mauller commented on GeneralsMD/Code/GameEngine/Source/GameClient/InGameUI.cpp:
```diff
@@ -3465,6 +3518,67 @@ void InGameUI::disregardDrawable( Drawable *draw )
 void InGameUI::postDraw( void )
```
Yes like all the other UI textual elements, it means they draw on top of everything else.

Mauller commented on GeneralsMD/Code/GameEngine/Source/GameClient/InGameUI.cpp:
```diff
@@ -3465,6 +3518,67 @@ void InGameUI::disregardDrawable( Drawable *draw )
+
+		// TheSuperHackers @info this implicitly offsets the simulation timer from the right instead of left of the screen
+		int horizontalTimerOffset = TheDisplay->getWidth() - (Int)m_simulationTimerPosition.x - m_simulationTimerString->getWidth() - m_simulationTimerFrameString->getWidth();
+		int horizontalFrameOffset = TheDisplay->getWidth() - (Int)m_simulationTimerPosition.x - m_simulationTimerFrameString->getWidth();
```
a bit shorter now that simulation is replaced with game

Mauller commented:
Made an update covering most points.

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/OptionsMenu.cpp:
```diff
@@ -772,6 +772,32 @@ Real OptionPreferences::getMusicVolume(void)
 	return volume;
 }
 
+// TheSuperHackers @info Here for ini parsing and future use when the options menu is reworked to add new options
```
We are missing records for these in `saveOptions`

```cpp
(*pref)["ShowSystemTime"] = TheWritableGlobalData->m_showSystemTime ? "yes" : "no";
```

Otherwise these options are not generated in the Options.ini

Mauller replied:
i didn't initially add them since there are no options menu features to toggling them, but i can block out the place for them and just save whatever the current value is.

xezon replied:
Yes, what this allows the user to do is edit these options manually in the ini file.

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/InGameUI.cpp:
```diff
@@ -876,6 +876,21 @@ const FieldParse InGameUI::s_fieldParseTable[] =
 
+	// TheSuperHackers @info ui enhancement configuration
+	{ "SystemTimeFont",					INI::parseAsciiString,	NULL, offsetof( InGameUI, m_systemTimeFont ) },
+	{ "SystemTimePointSize",			INI::parseInt,			NULL, offsetof( InGameUI, m_systemTimePointSize ) },
```
I think the point size can go into the Options.ini. GenTool also allows to customize the size.

And then we do not need the `ShowSystemTime` option, because setting `SystemTimePointSize` to 0 already implies that it is not shown.

Mauller replied:
Tweaked it so that the font size is controlled in the options and removed adjusting it from the ingameui.ini

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/InGameUI.cpp:
```diff
@@ -3465,6 +3518,64 @@ void InGameUI::disregardDrawable( Drawable *draw )
 {
 
+	// TheSuperHackers @info render the system time and game timer first, this way other text will render over them if they overlap
+	if (TheGlobalData->m_showSystemTime && !TheGameLogic->isInShellGame())
```
Clock can show everywhere. GenTool shows it always.

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/InGameUI.cpp:
```diff
@@ -1001,6 +1016,25 @@ InGameUI::InGameUI()
+	m_systemTimePointSize = 8;
+	m_systemTimeBold = TRUE;
+	m_systemTimePosition.x = 5; // TheSuperHackers @info relative to the left of the screen
+	m_systemTimePosition.y = 0;
```
GenTool has these values at x=3 and y=-1

On small resolution the x=5 looks a bit too large:

![shot_20250629_105231_1](https://github.com/user-attachments/assets/12a0fc4b-fea2-4cb1-90cb-5ba8894ffcb2)

Mauller replied:
Tweaked both to be 3 away from their respective side of the screen and 1 extra down from the top.
Y is positive down the screen in this instance.

![image](https://github.com/user-attachments/assets/e590e993-ad97-496f-b032-f364771f59ee)

Mauller replied:
i would say it looks better without the 1 down.

Mauller replied:
this is with the 3 and -1
![image](https://github.com/user-attachments/assets/272b008c-e166-4de4-8b4e-093b7d2f02a3)

Mauller replied:
> Isn't code for some of this stuff already there but only used for debug cheats?

That code is quite different to this and used for different purposes.

Mauller commented on GeneralsMD/Code/GameEngine/Source/GameClient/InGameUI.cpp:
```diff
@@ -876,6 +876,21 @@ const FieldParse InGameUI::s_fieldParseTable[] =
+	{ "SystemTimeBold",					INI::parseBool,			NULL, offsetof( InGameUI, m_systemTimeBold ) },
+	{ "SystemTimePosition",				INI::parseCoord2D,		NULL, offsetof( InGameUI, m_systemTimePosition ) },
+	{ "SystemTimeColor",				INI::parseColorInt,		NULL, offsetof( InGameUI, m_systemTimeColor ) },
+	{ "SystemTimeDropColor",			INI::parseColorInt,		NULL, offsetof( InGameUI, m_systemTimeDropColor ) },
```
i replicated what they were already doing, but switched it to spaces.

Mauller commented on GeneralsMD/Code/GameEngine/Source/GameClient/InGameUI.cpp:
```diff
@@ -3465,6 +3518,64 @@ void InGameUI::disregardDrawable( Drawable *draw )
 void InGameUI::postDraw( void )
 {
 
+	// TheSuperHackers @info render the system time and game timer first, this way other text will render over them if they overlap
```
i could do that, a followup PR could be done to then cleanup the other onscreen elements into their own functions too.

Mauller commented on GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/OptionsMenu.cpp:
```diff
@@ -772,6 +772,32 @@ Real OptionPreferences::getMusicVolume(void)
+	return FALSE;
+}
+
+// TheSuperHackers @info Here for ini parsing and future use when the options menu is reworked to add new options
```
i got rid of them, they aren't really needed to be honest.

Mauller commented:
Reworked based on recent suggestions

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/InGameUI.cpp:
```diff
@@ -3465,6 +3513,17 @@ void InGameUI::disregardDrawable( Drawable *draw )
 void InGameUI::postDraw( void )
```
Please test this with -quickstart

Chances are, clock is overdrawn by GUI.

I was experimenting with text drawing and figured we need a `postWindowDraw()` function in InGameUI. This function will be called after `TheWindowManager->winRepaint()`

Mauller replied:
So with quickstart the ingamegui is not running so the clock doesn't get rendered.

xezon replied:
I think it is running but you need to fix the draw order as highlighted above.

Mauller replied:
Did that and it will draw over menus now
![image](https://github.com/user-attachments/assets/a4a633fe-e36a-4d58-9dfc-142e4bc7ab5a)

Only thing is that it will also draw over other text that was draw in earlier draw steps.

xezon replied:
Yes it essentially will draw over everything. But it must do that. GenTool also did that. It is fine.

xezon commented on GeneralsMD/Code/GameEngine/Include/GameClient/InGameUI.h:
```diff
@@ -741,6 +746,25 @@ friend class Drawable;	// for selection/deselection transactions
+	Color												m_systemTimeColor;
+	Color												m_systemTimeDropColor;
+
+	// Simulation Timer
```
Maybe also call this "Game Time" here

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/OptionsMenu.cpp:
```diff
@@ -1262,6 +1290,26 @@ static void saveOptions( void )
 
+	//-------------------------------------------------------------------------------------------------
+	// Set System Time Font Size
+	val = TheWritableGlobalData->m_systemTimeFontSize; // TheSuperHackers @info replace with options input when applicable
```
`@todo`

2x

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/InGameUI.cpp:
```diff
@@ -3459,12 +3507,29 @@ void InGameUI::disregardDrawable( Drawable *draw )
 void InGameUI::postDraw( void )
 {
 
+	// TheSuperHackers @info render the game time first, this way other text will render over them if they overlap
```
I think this is unexpected. If System Time draws over everything then this can too. Given GenTool already does that since 15 years, players likely will expect that as well.

Mauller replied:
Moved the game time into the `postWindowDraw¬ function with the system time.

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/InGameUI.cpp:
```diff
@@ -5718,4 +5783,59 @@ WindowMsgHandledType IdleWorkerSystem( GameWindow *window, UnsignedInt msg,
+    TimeString.format(L"%2.2d:%2.2d:%2.2d", systemTime.wHour, systemTime.wMinute, systemTime.wSecond);
+	Int adjustedSystemTimeFontSize = TheGlobalLanguageData->adjustFontSize(m_systemTimePointSize);
+	GameFont* systemTimeFont = TheWindowManager->winFindFont(m_systemTimeFont, adjustedSystemTimeFontSize, m_systemTimeBold);
+	m_systemTimeString->setFont(systemTimeFont);
```
This setup is done every frame. This can be optimized by moving display string creation and setup to the constructor or init function. Can create a pair to `freeCustomUiResources`.

Mauller replied:
If we want people to be able to change the font size dynamically from the options menu then it will need to be setup every frame.

xezon replied:
The settings in the Options Menu are applied when pressing ACCEPT, which calls a function. In this function we will then be able to call

```cpp
freeCustomUiResources();
createCustomUiResources();
```

or

```cpp
refreshCustomUiResources();
```

to rescale the texts.

This would then also need to be done after the Display Resolution has changed.

xezon replied:
> I could also add in the FPS counter now if we wanted since the viewport scrolling changes got merged.

You can do this in a follow up change. We also need to add Network latency value. It is a value somewhere in network.

Mauller replied:
I got this working in the end. will push the changes soon.

xezon reviewed (commented):
Some more small things.

Mauller commented:
rebased with recent main

Mauller commented on GeneralsMD/Code/GameEngine/Source/GameClient/GameClient.cpp:
```diff
@@ -456,6 +456,8 @@ void GameClient::reset( void )
 	// need to reset the in game UI to clear drawables before they are destroyed
 	TheInGameUI->reset();
+	// TheSuperHackers @info we need to refresh the custom ui resources as their dependencies are cleared in the prior reset call
+	TheInGameUI->refreshCustomUiResources();
```
Yeah i did it that way at first since i had problems earlier on with display strting handling and the display string manager complaining.

But i stuck the refresh ui resources function in place of freeing the resources in ingameui::reset and it appears to work okay.

Mauller commented:
Tweaked and pushed, will replicate to generals soon.

Mauller commented:
Replicated to generals
